### PR TITLE
[Attention] Add direct symmetric-memory DCP A2A

### PR DESCRIPTION
**Pull request by GirasoleY**
## Context

Add an opt-in direct symmetric-memory A2A path for MLA DCP. With `VLLM_USE_DIRECT_DCP_A2A=1` and
`--dcp-comm-backend a2a`, each rank dispatches its partial output and LSE straight into peer receive buffers, peer then waits and performs the LSE-weighted combine on the receive buffer.

* This removes the stock pack -> NCCL all-to-all -> unpack/combine path from the hot operation. 
* The workspace is allocated once and shared across MLAlayers, with separate staging state per DBO ubatch. 

### Duplicate Check

| | This PR | #48248 |
|---|---|---|
| Exchange | One-sided writes directly into peer output/LSE buffers | FlashInfer LL128 all-to-all |
| Surrounding work | Eliminates stock pack/unpack and performs wait + LSE combine in CUDA | Retains pack/unpack and Triton combine |
| Runtime dependency | PyTorch symmetric memory in the local NVLink/NVSwitch domain | FlashInfer MNNVL workspace |
| Selection | Explicit opt-in; eager and graph-replay paths supported | Automatic Blackwell + full-CUDA-graph selection |

This PR is an alternative of the full a2a lse reduce path, can be merged independently of #48248.

### A2A + LSE-reduce latency reduction

Four GB200 GPUs, Kimi-K2.5 TP4/DCP4 geometry (64 total heads, 16 heads per
rank, head dimension 512, BF16). Each CUDA Graph contains 10 operations;
numbers are the median of seven repeats, with 100 graph replays per repeat,
reported as the maximum rank latency per operation. `stock` is the complete
current `dcp_a2a_lse_reduce` operation, including pack, NCCL A2A, and combine;
`direct` is the complete direct dispatch, signal, wait, and combine operation.

| Batch tokens | Stock (us) | Direct (us) | Latency reduction | Speedup |
|---:|---:|---:|---:|---:|
| 1 | 18.776 | 12.504 | 33.4% | 1.50x |
| 8 | 20.944 | 12.711 | 39.3% | 1.65x |
| 16 | 23.944 | 12.947 | 45.9% | 1.85x |
| 32 | 26.203 | 14.369 | 45.2% | 1.82x |
| 64 | 31.369 | 18.059 | 42.4% | 1.74x |
| 128 | 42.327 | 29.098 | 31.3% | 1.45x |

### InferenceX AgentX

Kimi-K2.5 DCP4, FlashInfer MLA AgentX

| Concurrency | Per-GPU tok/s, stock -> direct | Total gain | Output tok/s gain | Request/s gain | ITL p50 | ITL p95 |
|---:|---:|---:|---:|---:|---:|---:|
| 8 | 5,734 -> 5,778 | +0.77% | +0.20% | +0.56% | -4.05% | +0.57% |
| 32 | 8,160 -> 8,341 | +2.22% | +7.20% | +1.42% | -3.99% | +12.57% |
| 64 | 10,421 -> 11,164 | +7.14% | +4.45% | +4.71% | -5.86% | +18.80% |

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -2002,6 +2002,9 @@ def _resolve_rust_frontend_path() -> str | None:
     "VLLM_USE_SIMPLE_KV_OFFLOAD": lambda: bool(
         int(os.getenv("VLLM_USE_SIMPLE_KV_OFFLOAD", "0"))
     ),
+    "VLLM_USE_DIRECT_DCP_A2A": lambda: bool(
+        int(os.getenv("VLLM_USE_DIRECT_DCP_A2A", "0"))
+    ),
```
Comment (pavanimajety) [inline vllm/envs.py]: Why not use it by default?

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 2]
```diff
@@ -912,6 +937,18 @@ def process_weights_after_loading(self, act_dtype: torch.dtype):
...
+                get_dcp_group().device_group,
+                kv_b_proj_weight.device,
+                self.dcp_direct_a2a_max_num_tokens,
+                self.num_heads,
+                self.kv_lora_rank,
+                act_dtype,
+                self.dcp_direct_a2a_num_ubatches,
+            )
+            logger.info_once("Using direct symmetric-memory DCP A2A for MLA.")
+
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/mla_attention.py]: nit: why can't this be done at init time?

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 3]
```diff
@@ -845,7 +865,12 @@ def forward_impl(
 
             # correct dcp attn_out with lse.
             if self.impl.dcp_world_size > 1:
-                if self.dcp_a2a:
+                assert lse is not None
+                if self.dcp_direct_a2a_workspace is not None:
+                    attn_out = self.dcp_direct_a2a_workspace.lse_reduce(
+                        attn_out, lse, self.impl.lse_base_on_e
+                    )
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/mla_attention.py]: nit: i think id be cleaner/more-robust if it was 

```
if self.dcp_direct_a2a:
         assert self.dcp_direct_a2a_workspace is not None
         attn_out = self.dcp_direct_a2a_workspace.lse_reduce(
              attn_out, lse, self.impl.lse_base_on_e
          )
```

File: vllm/model_executor/layers/attention/mla_attention.py
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/mla_attention.py:1]: I think it might be worth refactoring this now so all the dispatching happens at init time (just a bit cleaner less booleans to track)

I think `MLAAttention` can take an explicit `device: torch.device`; `MultiHeadLatentAttentionWrapper` can pass `self.kv_b_proj.weight.device`. That lets the direct workspace be created in `__init__`, while the activation dtype is already available from the default dtype / model config:

```python
self.dcp_combine: Callable[
    [torch.Tensor, torch.Tensor], torch.Tensor
] | None = None

if self.dcp_direct_a2a:
    self. dcp_direct_a2a_workspace =  get_direct_dcp_a2a_workspace(
            group=get_dcp_group().device_group,
            device=device,
            max_num_tokens=min(
                vllm_config.scheduler_config.max_num_batched_tokens,
                max(
                    vllm_config.scheduler_config.max_num_seqs
                    * (1 + vllm_config.num_speculative_tokens),
                    vllm_config.compilation_config.max_cudagraph_capture_size or 0,
                ),
            heads_per_rank=self.num_heads,
            head_dim=self.kv_lora_rank,
            dtype=dtype,
            num_ubatches=vllm_config.parallel_config.num_ubatches,
     )
    self.dcp_combine = functools.partial(
        self. dcp_direct_a2a_workspace.lse_reduce,
        is_lse_base_on_e=self.impl.lse_base_on_e,
    )
elif self.impl.dcp_world_size > 1:
    dcp_combine = dcp_a2a_lse_reduce if self.dcp_a2a else cp_lse_ag_out_rs
    self.dcp_combine = functools.partial(
        dcp_combine,
        cp_group=get_dcp_group(),
        is_lse_base_on_e=self.impl.lse_base_on_e,
    )
```

The hot path then becomes:

```python
assert lse is not None
assert self.dcp_combine is not None
attn_out = self.dcp_combine(attn_out, lse)
```

thoughts?

**Review by LucasWilkinson — APPROVED** [review]
Overall looks pretty good to me; can this be on by default on CUDA devices? Ideally when we add new kernels we want them on by default in atleast one deployment scenario (can be HW/model/parallelism specific) just so the code is a bit more self documenting on why a kernel was added

File: vllm/v1/attention/ops/dcp_direct_a2a.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            symm_mem, signal_shape, torch.int32, device
+        )
+        self.epoch = torch.zeros(num_ubatches, dtype=torch.int64, device=device)
+
+    def _allocate(self, symm_mem, shape, dtype, device):
+        storage = symm_mem.empty(shape, device=device, dtype=dtype)
+        storage.zero_()
+        torch.accelerator.synchronize()
+        handle = symm_mem.rendezvous(storage, self.group.group_name)
+        handle.barrier()
```
Comment (pavanimajety) [inline vllm/v1/attention/ops/dcp_direct_a2a.py:90]:
```suggestion
        assert handle is not None, "DCP Symmetric Memory rendezvous returned None"
        handle.barrier()
```

**Review by pavanimajety — COMMENTED** [review]
I am currently running into an issue that returns handle as None. 
Do we need to set symmetric memory backend to NCCL explicitly? 
Could you share your vLLM server launch command?

File: vllm/model_executor/layers/attention/mla_attention.py (hunk 2 above)
Comment (GirasoleY) [reply vllm/model_executor/layers/attention/mla_attention.py]: addressed in 16913224

File: vllm/model_executor/layers/attention/mla_attention.py (hunk 3 above)
Comment (GirasoleY) [reply vllm/model_executor/layers/attention/mla_attention.py]: addressed in 16913224

File: vllm/model_executor/layers/attention/mla_attention.py
Comment (GirasoleY) [reply vllm/model_executor/layers/attention/mla_attention.py:1]: addressed in 16913224

File: vllm/v1/attention/ops/dcp_direct_a2a.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        num_ubatches: int = 1,
+    ) -> None:
+        import torch.distributed._symmetric_memory as symm_mem
+
+        if dtype not in (torch.float16, torch.bfloat16):
+            raise ValueError(f"Direct DCP A2A does not support {dtype}")
+        self.group = group
+        self.world_size = group.size()
+        self.rank = group.rank()
+        self.num_ubatches = num_ubatches
```
Comment (pavanimajety) [inline vllm/v1/attention/ops/dcp_direct_a2a.py:51]: Found the error: `num_ubatches` is set to 0 when dual batch optimization is not explicitly set. We need to throw an error here when `num_ubatches` are zero.

File: vllm/v1/attention/ops/dcp_direct_a2a.py (hunk 4 above)
Comment (GirasoleY) [reply vllm/v1/attention/ops/dcp_direct_a2a.py:90]: Thanks for catching this! My previous test had `--enable-dbo` set, 983f562 should fix the handle none issue. retesting it now.

File: vllm/envs.py (hunk 1 above)
Comment (GirasoleY) [reply vllm/envs.py]: This is now default on.

**Comment by GirasoleY** [comment]
> Overall looks pretty good to me; can this be on by default on CUDA devices? Ideally when we add new kernels we want them on by default in atleast one deployment scenario (can be HW/model/parallelism specific) just so the code is a bit more self documenting on why a kernel was added

The mode is default on with `VLLM_USE_DIRECT_DCP_A2A` unset. Set = 0 will explicit disable this.

**Review by pavanimajety — APPROVED** [review]
Validated the accuracy on my end, lgtm. Thanks for the PR!

**Comment by GirasoleY** [comment]
Feature is merged in https://github.com/vllm-project/vllm/pull/50484, closing this.
